support for `with socket.create_connection`

alicebob opened the pull request:
That would solve #1391 

Support for `with socket.create_connection` was added here: https://bugs.python.org/issue9794 in commit b383dbb45e62054e3a73792b7e903178bee18aa8 in cpython.

JelleZijlstra commented on stdlib/2and3/socket.pyi:
```diff
@@ -505,6 +505,10 @@ class socket:
                      proto: int = ..., fileno: Optional[int] = ...) -> None: ...
 
+    if sys.version_info >= (3, 2):
+        def __enter__(self) -> socket: ...
```
Strictly speaking this should be `def __enter__(self: _SelfT) -> _SelfT: ...`, where `_SelfT` is a TypeVar bound to socket. This way, the right type will be inferred if somebody uses `with` on a subclass of socket. Not sure how common it is to subclass `socket.socket`, but better to be safe.

alicebob commented:
Added that. Tested by hand (don't how where to add tests for this).

The new `_SelfT` needs the `socket` class, so I added it just where it's needed. If this is fine I'll squash it.

JelleZijlstra commented:
Thanks! Looks like you have a small lint error and pytype doesn't like defining typevars in the class scope; could you fix that?

alicebob commented:
Now it's the same as in stdlib/3/nntplib.py, which is the only other place I could find which defines `__enter__` with proper classes.

alicebob commented:
Build is happy, @JelleZijlstra

JelleZijlstra commented:
Thanks!

alicebob commented:
Thanks for the help. Would you be interested in a PR which adds typevars to more stdlib __enter__ methods?

JelleZijlstra commented:
Yes, that would be an improvement. It's not necessarily the highest priority thing though.
